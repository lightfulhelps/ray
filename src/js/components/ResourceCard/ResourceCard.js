// @flow
import * as React from 'react';
import classNames from 'classnames';
import { Icon } from '../../';

type Props = {
  campaign: string,
  className?: string,
  gradientStart?: string,
  gradientEnd?: string,
  resource: string,
  title: string,
};

const ResourceCard = ({
  campaign,
  className,
  gradientStart = '#27B0CC',
  gradientEnd = '#4513BB',
  resource,
  title,
  ...other
}: Props) => {
  const classes = classNames(
    className,
    'resource-card card p-1 rounded shadow h-100 d-flex flex-column justify-content-between'
  );

  return (
    <div
      {...other}
      className={classes}
      style={{
        backgroundImage: `linear-gradient(90deg, ${gradientStart} 0%, ${gradientEnd} 100%)`,
      }}
    >
      <div className="d-flex align-items-center mb-2">
        <div
          className="rounded-circle bg-gray-300 d-flex align-items-center justify-content-center mr-1"
          style={{ width: '24px', height: '24px' }}
        >
          <Icon name="chevronDown" theme="white" />
        </div>
        <div className="text-xs text-white">{campaign}</div>
      </div>
      <div className="text-sm font-weight-bold text-white">{title}</div>
    </div>
  );
};

export default ResourceCard;
