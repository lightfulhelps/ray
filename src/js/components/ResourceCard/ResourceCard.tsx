import * as React from 'react';
import classNames from 'classnames';
import { Icon } from '../..';

type Props = {
  campaign: string;
  className?: string;
  gradientEnd?: string;
  gradientStart?: string;
  imageForBackground?: string;
  resource: string;
  tag?: keyof JSX.IntrinsicElements;
  title: string;
};

export const isHex = (string: string): boolean => /^#([0-9A-F]{3}){1,2}$/i.test(string);

const ResourceCard: React.FC<Props> = ({
  campaign,
  className,
  gradientEnd,
  gradientStart,
  tag: Tag = 'a',
  title,
  imageForBackground,
  ...other
}: Props) => {
  const classes = classNames(
    className,
    'resource-card',
    'card p-1 rounded shadow h-100 bg-gray-900 text-decoration-none'
  );
  const style: { [key: string]: string | number } = {};

  if (imageForBackground) {
    style.backgroundImage = `url('${imageForBackground}')`;
    style.backgroundPosition = 'center';
    style.backgroundSize = 'cover';
    style.backgroundRepeat = 'no-repeat';
  } else if (gradientStart && isHex(gradientStart) && gradientEnd && isHex(gradientEnd)) {
    style.backgroundImage = `linear-gradient(90deg, ${gradientStart} 0%, ${gradientEnd} 100%)`;
  }

  return (
    <Tag {...other} className={classes} style={style}>
      <div className="d-flex align-items-center mb-2">
        <div
          className="rounded-circle d-flex align-items-center justify-content-center mr-1 flex-shrink-0"
          style={{ width: '24px', height: '24px', background: 'rgba(255, 255, 255, 0.2)' }}
        >
          <Icon isActive name="download" theme="white" />
        </div>
        <div className="text-xs text-white" data-test-id="resource-card-campaign">
          {campaign}
        </div>
      </div>
      <div className="text-sm font-weight-bold text-white" data-test-id="resource-card-title">
        {title}
      </div>
    </Tag>
  );
};

export default ResourceCard;
