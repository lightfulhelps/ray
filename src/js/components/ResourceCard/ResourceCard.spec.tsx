import React from "react";
import { shallow } from "enzyme";
import merge from "lodash/merge";
import ResourceCard, { isHex } from "./ResourceCard";

const setup = (overrides = {}) => {
  const props = merge({
    campaign: 'Fireside Storytelling',
    gradientStart: '#fafafa',
    gradientEnd: '#000000',
    resource: 'http://foo.com/bar.pdf',
    title: 'Useful storytelling guide'
  }, overrides);
  const wrapper = shallow(<ResourceCard {...props} />);

  return { wrapper, props };
};

describe('<ResourceCard />', () => {
  it('should render', () => {
    const {
      wrapper
    } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle className', () => {
    const {
      wrapper
    } = setup();

    wrapper.setProps({ className: 'custom' });

    expect(wrapper.hasClass('resource-card')).toBe(true);
    expect(wrapper.hasClass('custom')).toBe(true);
  });

  it('should handle campaign', () => {
    const campaign = 'My campaign';
    const {
      wrapper
    } = setup({ campaign });

    expect(wrapper.find('[data-test-id="resource-card-campaign"]').text()).toEqual(campaign);
  });

  it('should handle title', () => {
    const title = 'My title';
    const {
      wrapper
    } = setup({ title });

    expect(wrapper.find('[data-test-id="resource-card-title"]').text()).toEqual(title);
  });

  it('should not have a gradient background if gradientStart is an invalid HEX code', () => {
    const {
      wrapper
    } = setup({ gradientStart: 'foo', gradientEnd: '#ff0000' });

    expect(wrapper.prop('style').backgroundImage).toBeUndefined();
  });

  it('should not have a gradient background if gradientEnd is an invalid HEX code', () => {
    const {
      wrapper
    } = setup({ gradientStart: '#ff0000', gradientEnd: 'foo' });

    expect(wrapper.prop('style').backgroundImage).toBeUndefined();
  });

  it('should have a gradient background if gradientStart and gradientEnd are valid HEX codes', () => {
    const {
      wrapper
    } = setup({ gradientStart: '#ff0000', gradientEnd: '#fafafa' });

    expect(wrapper.prop('style').backgroundImage).toEqual('linear-gradient(90deg, #ff0000 0%, #fafafa 100%)');
  });

  it('should handle the tag prop', () => {
    const {
      wrapper
    } = setup();

    expect(wrapper.type()).toBe('a');

    wrapper.setProps({ tag: 'div' });

    expect(wrapper.type()).toBe('div');
  });

  it('should pass through other props', () => {
    const {
      wrapper
    } = setup({ tabIndex: 1, id: 'test' });

    expect(wrapper.prop('tabIndex')).toEqual(1);
    expect(wrapper.prop('id')).toEqual('test');
  });
});

describe('isHex', () => {
  it('should return true for hex-like strings', () => {
    expect(isHex('#000')).toBe(true);
    expect(isHex('#FFF')).toBe(true);
    expect(isHex('#fff')).toBe(true);
    expect(isHex('#FF0000')).toBe(true);
    expect(isHex('#ff0000')).toBe(true);
    expect(isHex('#FAFAFA')).toBe(true);
    expect(isHex('#fafafa')).toBe(true);
  });

  it('should return false for non hex-like strings', () => {
    expect(isHex('foo')).toBe(false);
    expect(isHex('000')).toBe(false);
    expect(isHex('FFF')).toBe(false);
    expect(isHex('#f')).toBe(false);
    expect(isHex('#ff')).toBe(false);
    expect(isHex('#ffff')).toBe(false);
    expect(isHex('#fffff')).toBe(false);
    expect(isHex('#fffffff')).toBe(false);
  });
});