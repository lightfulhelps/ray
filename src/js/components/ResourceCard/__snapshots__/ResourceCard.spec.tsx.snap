// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResourceCard /> should render 1`] = `
<a
  className="resource-card card p-1 rounded shadow h-100 bg-gray-900 text-decoration-none"
  resource="http://foo.com/bar.pdf"
  style={
    Object {
      "backgroundImage": "linear-gradient(90deg, #fafafa 0%, #000000 100%)",
    }
  }
>
  <div
    className="d-flex align-items-center mb-3"
  >
    <div
      className="rounded-circle d-flex align-items-center justify-content-center me-1 flex-shrink-0"
      style={
        Object {
          "background": "rgba(255, 255, 255, 0.2)",
          "height": "24px",
          "width": "24px",
        }
      }
    >
      <Icon
        isActive={true}
        name="download"
        theme="white"
      />
    </div>
    <div
      className="text-xs text-white"
      data-test-id="resource-card-campaign"
    >
      Fireside Storytelling
    </div>
  </div>
  <div
    className="text-sm fw-bold text-white"
    data-test-id="resource-card-title"
  >
    Useful storytelling guide
  </div>
</a>
`;
